### PR TITLE
[MRG] Add dPLI connectivity measure

### PR DESCRIPTION
**Pull request by kenjimarshall**
PR Description
--------------

Closes: #77 

Implementation of dPLI added to the `spectral_connectivity_epochs` module.

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [ ] [whats_new.rst](https://github.com/mne-tools/mne-connectivity/blob/main/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"

### CODE REVIEW

**Comment by kenjimarshall** [comment]
Hi @adam2392 yes I'll work on the example! And update the `whats_new.rst`. 

As per the unit tests, all the errors are from `test_spectral_connectivity_time_resolved` and `test_time_resolved_spectral_conn_regression` whenever the multitaper method is used. In particular, a shape mismatch keeps getting raised at line 219 in `time.py`.

E.g.

```
___________ test_spectral_connectivity_time_resolved[multitaper-coh] ___________
mne_connectivity/spectral/tests/test_spectral.py:476: in test_spectral_connectivity_time_resolved
    data, freqs=freqs, method=method, mode=mode)
mne_connectivity/spectral/time.py:219: in spectral_connectivity_time
    conn[epoch_idx, ...] = np.stack(conn_tr, axis=1)
E   ValueError: shape mismatch: value array of shape (2,6,3,18,256) could not be broadcast to indexing result of shape (2,6,18,256)
```

I'm not sure if/how my changes are causing this, but I can take a closer look at it if you'd like.

**Comment by adam2392** [comment]
> As per the unit tests, all the errors are from `test_spectral_connectivity_time_resolved` and `test_time_resolved_spectral_conn_regression` whenever the multitaper method is used. In particular, a shape mismatch keeps getting raised at line 219 in `time.py`.
> ```
> ___________ test_spectral_connectivity_time_resolved[multitaper-coh] ___________
> mne_connectivity/spectral/tests/test_spectral.py:476: in test_spectral_connectivity_time_resolved
>     data, freqs=freqs, method=method, mode=mode)
> mne_connectivity/spectral/time.py:219: in spectral_connectivity_time
>     conn[epoch_idx, ...] = np.stack(conn_tr, axis=1)
> E   ValueError: shape mismatch: value array of shape (2,6,3,18,256) could not be broadcast to indexing result of shape (2,6,18,256)
> ```

Is dPLI a symmetric measure, or is it asymmetric because it is directed? If so, then it's because the data structure constructed assumes a symmetric array. I can take a look later

**Comment by kenjimarshall** [comment]
Yes dPLI is asymmetric, so that could be it. Let me know how I should approach this.

**Comment by adam2392** [comment]
Hmmm... weird. I cannot reproduce the error locally.

**Comment by kenjimarshall** [comment]
I just gave you write permissions on the fork. I will work on the example in the meantime.

**Comment by adam2392** [comment]
@kenjimarshall I believe I fixed the issue. It was unrelated to your additions. 

The spectral connectivity over time has some issues, and are being sorted out right now still in #73

**Comment by kenjimarshall** [comment]
Great thank you!

**Comment by kenjimarshall** [comment]
I'm getting some  warnings when I build the docs locally for `spectral_connectivity_time`

```
WARNING: [numpydoc] Validation warnings while processing docstring for 'mne_connectivity.spectral_connectivity_time':
  PR01: Parameters {'sm_kernel'} not documented
  PR02: Unknown parameters {'kernel'}
  PR09: Parameter "sm_freqs" description should finish with "."
```

I just made those minor changes to fix circleCI.

**Comment by adam2392** [comment]
@kenjimarshall I've merged in extra changes from main branch here. Ping me when you want me to review the example

**Comment by adam2392** [comment]
Hi @kenjimarshall any challenges finishing this off?

**Comment by kenjimarshall** [comment]
Hi @adam2392 -- the example is now included

File: mne_connectivity/spectral/time.py [hunk 1]
```diff
@@ -269,6 +279,10 @@ def _spectral_connectivity(data, method, kernel, foi_idx,
     # get the supported connectivity function
     conn_func = {'coh': _coh, 'plv': _plv, 'sxy': _cs}[method]
 
+    # XXX: averages over tapers if necessary, but I don't think this is correct
+    if out.ndim == 5:
+        out = np.mean(out, axis=2)
+
```
Comment (adam2392) [inline mne_connectivity/spectral/time.py:285]:
```suggestion
```

Comment (adam2392) [reply mne_connectivity/spectral/time.py:285]: Not needed because of below.

File: mne_connectivity/spectral/time.py [hunk 2]
```diff
@@ -215,8 +215,18 @@ def spectral_connectivity_time(data, names=None, method='coh', indices=None,
...
+        print(data.shape)
+        print(conn.shape)
+        print([x.shape for x in conn_tr])
+
+        # average across tapers if necessary
+        conn_tr = np.stack(conn_tr, axis=1)
+        if all(x.ndim == 4 for x in conn_tr):
+            conn_tr = np.mean(conn, axis=1)
+
+        print(conn_tr.shape)
```
Comment (adam2392) [inline mne_connectivity/spectral/time.py:227]:
```suggestion
```

File: mne_connectivity/spectral/time.py [hunk 3]
```diff
@@ -215,8 +215,18 @@ def spectral_connectivity_time(data, names=None, method='coh', indices=None,
...
         # merge results
-        conn[epoch_idx, ...] = np.stack(conn_tr, axis=1)
+        conn[epoch_idx, ...] = conn_tr
```
Comment (adam2392) [inline mne_connectivity/spectral/time.py:229]:
```suggestion
        conn[epoch_idx, ...] = np.stack(conn_tr, axis=1)
```

File: examples/dpli_wpli_pli.py [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+# :math:`k \in \mathbb{Z}`). This property provides protection against
+# recognizing volume conduction effects as connectivity, and is the backbone
+# for these methods :footcite:`VinckEtAl2011`.
+#
+# :math:`PLI = |E[sgn(\mathcal{I}(X_{ij}))]|` :footcite:`StamEtAl2007`
+#
+# :math:`wPLI = \frac{|E[\mathcal{I}(X_{ij})]|}{E[|\mathcal{I}(X_{ij})|]}`
+# :footcite:`VinckEtAl2011`
+#
+# :math:`dPLI = E[\mathcal{H}(\mathcal{I}(X_{ij}))]` :footcite:`StamEtAl2012`
```
Comment (adam2392) [inline examples/dpli_wpli_pli.py:44]: Perhaps it would be beneficial for the reader to define what $\mathcal{I}$ and $\mathcal{H}$ is and $sgn$.

**Comment by adam2392** [comment]
Great start and contribution @kenjimarshall ! I'm pinging some other folks here to take a look, since examples should always be pretty generally understandable by a new user/viewer. 

This might go through some reviews if you're okay with that. Just want to iterate to make sure the example is as good as possible before merging in.

**Comment by kenjimarshall** [comment]
Hi @adam2392, I just wanted to check in on this PR. Let me know if there are any other changes I can make while the example awaits further review.

**Comment by adam2392** [comment]
https://output.circle-artifacts.com/output/job/c5dd2ea3-543e-4542-9cd9-aad54a3459c6/artifacts/0/dev/auto_examples/dpli_wpli_pli.html#sphx-glr-auto-examples-dpli-wpli-pli-py

@agramfort @britta-wstnr @larsoner any thoughts?

File: doc/whats_new.rst [hunk 5]
```diff
@@ -39,7 +40,7 @@ API
 Authors
 ~~~~~~~
 
-
+* `Kenji Marshall`_
```
Comment (adam2392) [inline doc/whats_new.rst:43]:
```suggestion
* `Kenji Marshall`_
* `Adam Li`_
* `Alex Rockhill`_
* `Szonja Weigl`_
```

**Comment by adam2392** [comment]
Thanks @kenjimarshall !

**Pull request merged**